packages/core/data-transfer/src/directory/providers/source/index.ts

```diff
@@ -0,0 +1,208 @@
+      throw new ProviderInitializationError('Could not load metadata from the directory.');
+    }
+  }
+
+  async #loadMetadata() {
+    const metadataPath = path.join(this.options.directory.path, METADATA_FILE_PATH);
+    this.#metadata = await this.#readJSONFile<IMetadata>(metadataPath);
+  }
+
+  async #loadAssetMetadata(filePath: string) {
+    return this.#readJSONFile<IFile>(filePath);
+  }
+
+  async getMetadata() {
+    if (!this.#metadata) {
+      await this.#loadMetadata();
+    }
+
+    return this.#metadata ?? null;
+  }
+
+  async getSchemas() {
+    const schemaCollection = await utils.stream.collect<Struct.Schema>(
+      this.createSchemasReadStream()
+    );
+
+    if (isEmpty(schemaCollection)) {
+      throw new ProviderInitializationError('Could not load schemas from the directory.');
+    }
+
+    // Group schema by UID
+    const schemas = keyBy('uid', schemaCollection);
+
+    // Transform to valid JSON
+    return utils.schema.schemasToValidJSON(schemas);
+  }
+
+  createEntitiesReadStream(): Readable {
+    return this.#streamJsonlDirectory('entities');
+  }
+
+  createSchemasReadStream(): Readable {
+    return this.#streamJsonlDirectory('schemas');
+  }
+
+  createLinksReadStream(): Readable {
+    return this.#streamJsonlDirectory('links');
+  }
+
+  createConfigurationReadStream(): Readable {
+    return this.#streamJsonlDirectory('configuration');
+  }
+
+  createAssetsReadStream(): Readable {
+    const outStream = new PassThrough({ objectMode: true });
+    const assetsDir = path.join(this.options.directory.path, 'assets/uploads');
+
+    fs.readdir(assetsDir, { withFileTypes: true }, (err, files) => {
+      if (err) {
+        outStream.destroy(new ProviderTransferError(`Failed to read assets directory: ${err}`));
+        return;
+      }
+
+      files
+        .filter((file) => file.isFile())
+        .forEach((file) => {
+          const filePath = path.join(assetsDir, file.name);
+          const metadataPath = path.join(
+            this.options.directory.path,
+            'assets/metadata',
+            `${file.name}.json`
+          );
+          const size = fs.statSync(filePath).size;
+
+          let metadata;
+          try {
+            metadata = fs.readJSONSync(metadataPath);
+          } catch {
+            console.warn(`Failed to read metadata for ${file.name}`);
+          }
+
+          const asset: IAsset = {
+            metadata,
+            filename: file.name,
+            filepath: filePath,
+            stats: { size },
+            stream: fs.createReadStream(filePath),
+          };
+
+          outStream.write(asset);
+        });
+
+      outStream.end();
+    });
+
+    return outStream;
+  }
+
+  #streamJsonlDirectory(subDirectory: string): Readable {
+    const fullPath = path.join(this.options.directory.path, subDirectory);
+    const outStream = new PassThrough({ objectMode: true });
+
+    fs.readdir(fullPath, { withFileTypes: true }, (err, files) => {
```

Comment:
Same than for the assets read stream, maybe we could try to create a stream using an async iterator to avoid introducing potential issues (like oversights in our custom impl) + keep it easy to read?